### PR TITLE
Fix Security Monitoring flaky tests

### PR DESCRIPTION
**Pull request by MatiasDwek**
### What does this PR do?

Fix Security Monitoring API integration test flakiness.
These tests were flaky for the following reasons:
- After creating a rule, we need to wait some time before sending logs to that rule.
- We need retries each time we retrieve signals.
- We need to wait some time when sending logs to trigger signals if we want to check the order of the generated signals's timestamps.

### Additional Notes

Number of retries and waiting times are somewhat arbitrary. Might need some tweaking.

### Review checklist
Please check relevant items below:
- [ ] This PR includes all [newly recorded cassettes](/DEVELOPMENT.md) for any modified tests.

- [ ] This PR does not rely on API client schema changes.
    - [ ] The CI should be fully passing.
- [ ] Or, this PR relies on API schema changes and this is a Draft PR that includes tests.
  - Note: CI shouldn't be run on this Draft PR, as its expected to fail without the corresponding schema changes.

### CODE REVIEW

File: src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java [hunk 1]
```diff
@@ -242,71 +242,103 @@ public void searchSignals() throws Exception {
...
 
-        assertNotEquals(pageOneResponse.getData().get(0).getId(), pageTwoResponse.getData().get(0).getId());
+                boolean correctSecondLink = !pageOneResponse.getData().get(0).getId().equals(pageTwoResponse.getData().get(0).getId());
+                return correctFirstLink && correctSecondLink;
+            } catch (ApiException ignored) {
+                return false;
+            }
+        });
 
         api.deleteSecurityMonitoringRule(rule.getId());
```
Comment (MatiasDwek) [inline src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java:343]: If the test fails, we won't do this clean up step. Not sure how we can do a "finally" in Junit, besides @After

File: src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java [hunk 2]
```diff
@@ -192,121 +194,181 @@ public void searchSignals() throws Exception {
...
         });
 
-        // Sort works correctly
-        SecurityMonitoringSignalsListResponse responseAscending = api.searchSecurityMonitoringSignals()
-                .body(new SecurityMonitoringSignalListRequest()
-                        .filter(allSignalsFilter)
-                        .sort(SecurityMonitoringSignalsSort.TIMESTAMP_ASCENDING))
-                .execute();
+        // Sort ascending works correctly
+        TestUtils.retry(10, 10, () -> {
```
Comment (btoueg) [inline src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java:198]: Can't we have 5-10 everywhere?

File: src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java [hunk 3]
```diff
@@ -192,121 +194,181 @@ public void searchSignals() throws Exception {
...
+                        .body(new SecurityMonitoringSignalListRequest()
+                                .filter(allSignalsFilter)
+                                .page(new SecurityMonitoringSignalListRequestPage()
+                                        .cursor(cursor)
+                                        .limit(1)))
+                        .execute();
+                if (pageTwoResponse.getData() == null || pageTwoResponse.getData().size() < 1)
+                    return false;
+                
+                boolean correctSecondLink = !pageOneResponse.getData().get(0).getId().equals( pageTwoResponse.getData().get(0).getId());
```
Comment (btoueg) [inline src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java:262]: We are testing first link of first page is not the same as first link on second page. `correctSecondLink` is a confusing name IMO

File: src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java [hunk 4]
```diff
@@ -171,7 +171,9 @@ public void rulesCrudTest() throws ApiException {
     @Test
     public void searchSignals() throws Exception {
         String uniqueName = getUniqueEntityName();
-        SecurityMonitoringRuleResponse rule = createRule(uniqueName);
+        createRule(uniqueName);
+        // Wait for the rules to be created
```
Comment (greentruff) [inline src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java:175]:
```suggestion
        // Wait for the backend to start processing the rule
```

File: src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java [hunk 5]
```diff
@@ -192,121 +194,181 @@ public void searchSignals() throws Exception {
...
+                        .pageLimit(1)
+                        .pageCursor(cursor)
+                        .execute();
 
-        assertEquals(1, pageTwoResponse.getData().size());
+                if (pageTwoResponse.getData() == null || pageTwoResponse.getData().size() < 1)
+                    return false;
 
-        assertNotEquals(pageOneResponse.getData().get(0).getId(), pageTwoResponse.getData().get(0).getId());
+                boolean correctSecondLink = !pageOneResponse.getData().get(0).getId().equals(pageTwoResponse.getData().get(0).getId());
```
Comment (btoueg) [inline src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java:366]: Same

**Review by btoueg — APPROVED** [review]
LGTM

File: src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java [hunk 6]
```diff
@@ -192,121 +194,181 @@ public void searchSignals() throws Exception {
...
+                                .filter(allSignalsFilter)
+                                .sort(SecurityMonitoringSignalsSort.TIMESTAMP_ASCENDING))
+                        .execute();
 
-        assertEquals(2, responseAscending.getData().size());
-        OffsetDateTime firstTimestamp = responseAscending.getData().get(0).getAttributes().getTimestamp();
-        OffsetDateTime secondTimestamp = responseAscending.getData().get(1).getAttributes().getTimestamp();
-        assertTrue(firstTimestamp.isBefore(secondTimestamp) || firstTimestamp.isEqual(secondTimestamp));
+                if (responseAscending.getData() == null || responseAscending.getData().size() < 2)
+                    return false;
```
Comment (greentruff) [inline src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java:207]: Can you add brackets here?

File: src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java (hunk 1 above)
Comment (greentruff) [reply src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java:373]: The normal pattern is 
```
try {
} finally {
  // cleanup
}
```
`@After ` might be sufficient though

**Review by greentruff — APPROVED** [review]

File: src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java (hunk 6 above)
Comment (MatiasDwek) [reply src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java:207]: yes good catch, I fixed all of them, my IDE save actions were disabled

File: src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java (hunk 3 above)
Comment (MatiasDwek) [reply src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java]: indeed, I corrected it!

File: src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java [hunk 7]
```diff
@@ -171,7 +171,9 @@ public void rulesCrudTest() throws ApiException {
     @Test
     public void searchSignals() throws Exception {
         String uniqueName = getUniqueEntityName();
-        SecurityMonitoringRuleResponse rule = createRule(uniqueName);
+        createRule(uniqueName);
+        // Wait for the backend to start processing the rule
+        Thread.sleep(5000);
```
Comment (therve) [inline src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java:176]: Please skip those sleep calls when we use recorded cassettes.

Comment (MatiasDwek) [reply src/test/java/com/datadog/api/v2/client/api/SecurityMonitoringApiTest.java]: Fixed!

**Review by therve — APPROVED** [review]

**Pull request merged**